Utils/Dataflow/016_adjustMetadata/adjustMetadata.py

```diff
@@ -204,6 +184,10 @@ def transform_chain_data(data):
     chain_data = data.get('chain_data')
     if not chain_data or not chain_data.replace(',', '').isdigit():
         taskid = data.get('taskid')
+        if not taskid:
+            sys.stderr.write('(WARN) No taskid: %s. Skipping chain_data '
+                             'processing.\n')
```

Comment:
It's the latter. Turned out that it's not just a matter of adding the argument: 90c3014.

Reply:
About this commit:
> 90c3014

Please take a look here: https://trello.com/c/JVcdkaH3. If we fix this moment, wouldn\`t it help (to avoid storing original message artificially)?

Another moment: would\`t it be more logical to return errors back to `main()` and make *it* responsible for correct errors representation? Not sure how exactly it can be done, though: error message as return value, or adding special field `_error` to the data, (combined with return value `False`), ... Of course, special field in data won\`t work if passed parameter was not of type `dict`; but if it is the case -- then it does not matter what came to the whole stage, the only interesting thing is "what came to the given tranformation step?"

Or maybe simply agree that message ID is `data['taskid']`, and use it for all error messages; and to make sure there *is* some `taskid`, start all the processing with input message check at step zero. After all, if there\`s no `'taskid'` (or input message content is not of type `dict` at all), something definitely went wrong and most likely the message is going to be thrown away.

Reply:
> About this commit:
> 90c3014
> Please take a look here: https://trello.com/c/JVcdkaH3. If we fix this moment, wouldn`t it help (to avoid storing original message artificially)?

I don't think so.

If I understand the trello card correctly, then the problem it describes is the fact that the message's object (`decoded` in this case) is passed outside of the message, where it can be (unintentionally) changed "by hand", while it is supposed to be changed only by using the message's methods.

While this seems to be the case in this stage, the issue this commit is intended to solve is different: `data` (regardless of the way it was obtained) changes when it passes through a processing function, which makes finding the original message difficult unless it is saved explicitly (like done in this commit). Solving the first problem (by replacing `return self.decoded` with `return dict(self.decoded)`, I presume) won't affect the second one.

> Another moment: would`t it be more logical to return errors back to main() and make it responsible for correct errors representation? Not sure how exactly it can be done, though: error message as return value, or adding special field _error to the data, (combined with return value False), ... Of course, special field in data won`t work if passed parameter was not of type dict; but if it is the case -- then it does not matter what came to the whole stage, the only interesting thing is "what came to the given tranformation step?"

Yes, this looks like a good idea to me. For example, before separation of the stage 016 two functions were checking the existence of `taskid` (actually, the second one did not perform the check due to relying on the first one, which is also incorrect). Also, this should be done because functions can possibly be reused elsewhere - including code where errors should be reported in a different way.

I don't think that a special field is required (and can even be harmful in aforementioned reusage case), error code/message in the return value should do.

> Or maybe simply agree that message ID is data['taskid'], and use it for all error messages; and to make sure there is some taskid, start all the processing with input message check at step zero. After all, if there`s no 'taskid' (or input message content is not of type dict at all), something definitely went wrong and most likely the message is going to be thrown away.

This sounds reasonable as well. Will do.

Reply:
Update: fixing the first problem can actually help us if we decide to pass `msg.content()` instead of stored `original` to processing functions. However, in order for this to work the `content()` and `decode()` methods must be "cheaper" than storing `original`. When they simply pass the saved `dict` - sure, no problem, but I'm not sure if this will always be the case.

Reply:
@Evildoor:
>@mgolosova:
>>Or maybe simply agree that message ID is data['taskid'], and use it for all error messages; and to make sure there is some taskid, start all the processing with input message check at step zero. After all, if there`s no 'taskid' (or input message content is not of type dict at all), something definitely went wrong and most likely the message is going to be thrown away.
>
>This sounds reasonable as well. Will do.

Ok, thank you.

---

@Evildoor:
>@mgolosova:
>>Please take a look here: https://trello.com/c/JVcdkaH3. If we fix this moment, wouldn`t it help (to avoid storing original message artificially)?
>
>fixing the ... problem can actually help us if we decide to pass `msg.content()` instead of stored original to processing functions. However, in order for this to work the `content()` and `decode()` methods must be "cheaper" than storing original. When they simply pass the saved dict - sure, no problem, but I'm not sure if this will always be the case.

It is not about "cheaper in execution" (Python is not the "cheapest" language itself), but about "cheaper in coding". I mean, whenever we might need the access to the original message, we would have to have something like this line: `original = dict(data)` in the beginning of the processing, right after `data = msg.decode()` [1] -- just to make it possible. It makes me think about more regular way to achieve it. I believe that introducing a new variable worth doing when the it is used to store `result.of(some_repeating(long, expression)) + maybe(with, conditions())`. But here it stores `msg.content()` -- which is not that long and, I want to believe, quite transparent.

Still it was nothing but a comment, not to say it should be done here-and-now. Just as you stumbled over this thing once, I thought it might be interesting for you to glance at the related issue.

[1] ...which also makes me doubtful (maybe we should just pass already decoded message to `process()`?), but it is not the matter of this PR for sure.

Reply:
Aha, I see your point. Thanks for the explanation.